src/Agent.Listener/CommandSettings.cs

```diff
@@ -7,121 +7,84 @@
-            Constants.Agent.CommandLine.Args.StartupType,
-            Constants.Agent.CommandLine.Args.Token,
-            Constants.Agent.CommandLine.Args.Url,
-            Constants.Agent.CommandLine.Args.UserName,
-            Constants.Agent.CommandLine.Args.WindowsLogonAccount,
-            Constants.Agent.CommandLine.Args.WindowsLogonPassword,
-            Constants.Agent.CommandLine.Args.Work
-        };
+        // Commands
+        private ConfigureAgent Configure { get; set; }
+        private RemoveAgent Remove { get; set; }
+        private RunAgent Run { get; set; }
+        private WarmupAgent Warmup { get; set; }
+
+        public IEnumerable<Error> ParseErrors { get; set; }
 
-        // Commands.
-        public bool Configure => TestCommand(Constants.Agent.CommandLine.Commands.Configure);
-        public bool Remove => TestCommand(Constants.Agent.CommandLine.Commands.Remove);
-        public bool Run => TestCommand(Constants.Agent.CommandLine.Commands.Run);
-        public bool Warmup => TestCommand(Constants.Agent.CommandLine.Commands.Warmup);
-
-        // Flags.
-        public bool Commit => TestFlag(Constants.Agent.CommandLine.Flags.Commit);
-        public bool Diagnostics => TestFlag(Constants.Agent.CommandLine.Flags.Diagnostics);
-        public bool Help => TestFlag(Constants.Agent.CommandLine.Flags.Help);
-        public bool Unattended => TestFlag(Constants.Agent.CommandLine.Flags.Unattended);
-        public bool Version => TestFlag(Constants.Agent.CommandLine.Flags.Version);
-        public bool DeploymentGroup => TestFlag(Constants.Agent.CommandLine.Flags.MachineGroup) || TestFlag(Constants.Agent.CommandLine.Flags.DeploymentGroup);
-        public bool DeploymentPool => TestFlag(Constants.Agent.CommandLine.Flags.DeploymentPool);
-        public bool EnvironmentVMResource => TestFlag(Constants.Agent.CommandLine.Flags.Environment);
-        public bool GitUseSChannel => TestFlag(Constants.Agent.CommandLine.Flags.GitUseSChannel);
-        public bool RunOnce => TestFlag(Constants.Agent.CommandLine.Flags.Once);
 
         // Constructor.
-        public CommandSettings(IHostContext context, string[] args, IScopedEnvironment environmentScope=null)
+        public CommandSettings(IHostContext context, string[] args, IScopedEnvironment environmentScope = null)
         {
             ArgUtil.NotNull(context, nameof(context));
             _promptManager = context.GetService<IPromptManager>();
             _trace = context.GetTrace(nameof(CommandSettings));
 
-            // Parse the command line args.
-            _parser = new CommandLineParser(
-                hostContext: context,
-                secretArgNames: Constants.Agent.CommandLine.Args.Secrets);
-            _parser.Parse(args);
+
+            // Parse once to record Errors
+            ParseArguments(args, false);
+
+            // Parse a second time to populate objects (even if there are errors)
+            ParseArguments(args, true);
```

Comment:
I am not sure I understand why we need to parse arguments  twice. Is ParserArguments() our method?

Reply:
Check out the comment above.  There is an option to parse ignoring errors.  The way the old parser worked is that it ignored errors.  There is a solid argument that you shouldn't do that, however enforcing that now breaks backwards compat.

So we parse once, to get any errors.
We parse a second time to map ignoring errors.  

Now that you mention I can definitely add a check on the second call, to see if any errors occurred and only run it if necessary.  That would definitely be an improvement and make things clearer.

Reply:
would it make sense to have ```ParseArguments()``` itself detect if there were any errors, only then call itself again but this time set to ignore the errors?

Reply:
Yes I can make that happen :-)

Reply:
I cleaned this up.  I could have done a recursive call to fix but did it the way I did because I think it's easier to read.  Have a look and let me know what you think.